### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by meredydd**
This PR combines a series of performance improvements, totalling a ~45%~ **85%** speedup (aka a ~31%~ 46% reduction in runtime) on @rixner's student project ([`field.py`](https://gist.githubusercontent.com/rixner/f79dfa52a0385e5c82d147ae3d06c50c/raw/971a46b80c9ff79734e80f7bab49f66ecdc3ae11/field.py)).

Tagging @rixner because this is relevant to your interests, @albertjan because @rixner is busy, and @LeszekSwirski for arcane mysteries of V8 performance.

I'd suggest reviewing each commit individually: Each improvement is a single commit, with a paragraph of notes in the commit message.

Performance numbers throughout use the median of 10ish measurements on my laptop, of 500 ticks of `field.py`. (I made a desultory effort to get the CPU throttling into a consistent state, but it's pretty rough; T tests averaged around p=0.10.)

----
**Notes and Highlights:**

The hottest code was in the attribute-getting infrastructure:

* We win 16% by reworking an inefficient fallback from `klass.prototype.tp$getattr()` to `object.GenericPythonGetAttr`, which is of course the common case.

* Removing `klass.prototype.tp$getattr()` entirely when unnecessary (ie when the class doesn't define a custom `__getattribute__`) got us another 10%

* `Sk.abstr.gattr()` is an anomalously hot spot. Even after streamlining, it's still nearly 4% of CPU (according to the Chrome profiler), despite doing very little!

  - I initially tried inlining the `tp$getattr()` call straight into generated code, which won us quite a lot (4.5% speedup). I'm guessing this is because V8 could now do call-site-specific optimisation (each line of user code pretty consistently calls one `tp$getattr` implementation).

  - However, the benefit from that inlining basically disappeared after I removed `tp$getattr()` from most classes. This is presumably because, now that most classes share a single `tp$getattr()` implementation (in fact, _all_ classes in this particular program do), V8 can do call-site-specific optimisation on the `tp$getattr()` call in `Sk.abstr.gattr()` itself.

    I am slightly concerned that more complex programs that manipulate other types -- modules, exceptions, classes that implement `__getattribute__`, etc -- might prevent V8 from optimising that site. So perhaps we should still be inlining those calls, but I spent 5 minutes messing around with microbenchmarks and couldn't find evidence to justify it, so I'm holding off for now. (It's in this PR, but reverted.)

* Most attrs are looked up in the object dictionary, so a fast path for string comparisons in `Sk.builtin.dict.prototype.key$lookup` proved well worth it, netting us another 21% (harvested in two parts: 8.8% from remarkably small tweaks to `richCompareBool`, and then 16.3% from bypassing it altogether when doing string lookups)

The function-call infrastructure is next on my hit list. I've grabbed a couple of pieces of low-hanging fruit, but there's plenty still to get:

* `Sk.misceval.callsimOrSuspendArray()` has an API almost identical to `tp$call()`, but is much slower (it calls `applyOrSuspend()`, which goes round the houses of all the things Skulpt knows how to invoke). So I've added a fast path that invokes `tp$call` directly from generated code when possible.

  - I'm not actually sure we _should_ support invoking anything that doesn't have a `tp$call` method. We shouldn't sacrifice common-case performance to make things marginally easier for JS devs (eg allowing you to invoke JS functions). Forcing _all_ calls to use `tp$call` would speed up everything and clear out a bunch of code, but it would break our external API so I've held off.

* I've cherry-picked #950 into this, and the fast path is indeed catching most executions, but `func.tp$call` (and `method.tp$call`) are still taking ~4% of runtime each. Plenty to be getting on with here!

### CODE REVIEW

File: src/misceval.js [hunk 1]
```diff
@@ -884,14 +884,24 @@ Sk.exportSymbol("Sk.misceval.callsimOrSuspend", Sk.misceval.callsimOrSuspend);
...
-    return Sk.misceval.applyOrSuspend(func, undefined, undefined, undefined, argarray);
+Sk.misceval.callsimOrSuspendArray = function (func, args, kws) {
+    if (!args) {
+        args = [];
+    }
+    if (func.tp$call) {
+        return func.tp$call(args, kws);
+    } else {
+        // Slow path handles things like calling native JS fns
+        // (perhaps we should stop supporting that), and weird
```
Comment (albertjan) [inline src/misceval.js]: we should stop supporting raw js functions

**Comment by rixner** [comment]
@meredydd you are on fire here!  I will look at these in detail this week!

**Comment by rixner** [comment]
I am seeing a 2X speedup for the whole thing on a different benchmark.  Very nice work!  I'll look commit-by-commit.

File: src/dict.js [hunk 2]
```diff
@@ -76,6 +76,18 @@ Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
     var eq;
     var i;
 
+    // Fast path: We spend a *lot* of time looking up strings
+    // in dictionaries. (Every attribute access, for starters.)
+    if (key.ob$type === Sk.builtin.str) {
+        for (i = 0; i < bucket.items.length; i++) {
+            item = bucket.items[i];
+            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
```
Comment (rixner) [inline src/dict.js]: I marginally prefer using ```item.lhs.$jsstr() === key.$jsstr()``` here to get away from assuming ```.v``` is the representation, but this is internal and it is so pervasive, that it probably doesn't matter...

File: src/misceval.js (hunk 1 above)
Comment (rixner) [reply src/misceval.js]: I agree.  Do any still exist in the core of Skulpt?

File: src/type.js [hunk 3]
```diff
@@ -459,6 +441,30 @@ Sk.builtin.type = function (name, bases, dict) {
...
+                );
+                return canSuspend ? r : Sk.misceval.retryOptionalSuspensionOrThrow(r);
+            };
+        } else if (!klass.prototype.tp$getattr) {
+            // This is only relevant in Python 2, where
+            // it's possible not to inherit from object
+            // (or perhaps when inheriting from builtins? Unclear)
+            klass.prototype.tp$getattr = Sk.builtin.object.prototype.GenericGetAttr;
+        }
+
```
Comment (rixner) [inline src/type.js]: What happens if you monkey patch the class to add a ```__getatttribute__``` method here?  Personally, I don't care, because you deserve what you get (in my opinion!) if you do that, but does that create a correctness issue?

File: src/dict.js (hunk 2 above)
Comment (meredydd) [reply src/dict.js]: That function dispatch would be expensive (this code is red-hot); if there were _anywhere_ to inline `$jsstr()`, it's here.

File: src/misceval.js (hunk 1 above)
Comment (meredydd) [reply src/misceval.js]: I didn't spot any, but I'm not confident I turned over every rock.

File: src/type.js (hunk 3 above)
Comment (meredydd) [reply src/type.js]: You're right; it now breaks, like every other dunder-method (monkey-patching of `__len__`, `__getitem__`, or others isn't reflected in the JS slot functions either).

I can imagine a custom `tp$setattr` for classes that updates the prototype if you do a monkey-patch like that...although that won't work if you do it to a multiply-inherited ancestor (JS prototypical inheritance does not follow the Python MRO!). Subclasses could register a callback to be notified if it happens, I suppose, but that makes subclasses non-garbage-collectable, which I'm not too keen on.

Or we could set a global "someone monkey-patched something important" flag, which causes _all_ slot functions to deoptimise forevermore...although that doesn't help if the slot functions aren't even defined (which is the case for `sq$len` and friends).

Basically, it's a gnarly one to support, and our support for all the _other_ slot functions is already intolerant of monkey-patching, so I'm inclined to let it stay broken unless we can work out how to fix them all at once.

**Comment by meredydd** [comment]
> However, I would at least like you to think about/respond to the issue of adding a `__getattribute__` method to a class dynamically and whether it is actually a problem (and/or if you care about it).

See my reply to your comment. Yes, this breaks a piece of monkey-patching, but we don't support monkey-patching anywhere else in the dunder/slot functions, so I'm (reluctantly, and while holding my nose) calling it an acceptable cost.

If there were a reasonably performant way of making monkey patching actually work, I'd be up for it.

File: src/dict.js (hunk 2 above)
Comment (rixner) [reply src/dict.js]: Fair enough.

**Comment by rixner** [comment]
As I implied, I'm not a fan of monkey patching, so am not at all offended by this.  Especially since we don't fully support it elsewhere anyway.  This was a minor step backward in that regard, but the win is large and I agree with the sentiment that if we believe this should be supported, they all should support it, not just selected methods.  I'm going to very happily merge this!  Again: great work!
